### PR TITLE
nixos/hardware/cpu/x86/msr: init

### PR DESCRIPTION
**Pull request by lorenzleutgeb**
## Description of changes

Introduces a new module at `nixos/modules/hardware/cpu/x86-msr.nix` which adds four regular options and one freeform option:

 - `hardware.cpu.x86.msr.enable`
 - `hardware.cpu.x86.msr.group`
 - `hardware.cpu.x86.msr.mode`
 - `hardware.cpu.x86.msr.owner`
 - `hardware.cpu.x86.msr.settings` (see [Nix RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md))

If enabled, it will:

 1. Add "msr" to the list of kernel modules to be loaded (with the parameters from `hardware.cpu.x86.msr.settings`).
 2. Create the group "msr", if `hardware.cpu.x86.msr.group == "msr"`.
 3. Add an udev rule for devices that match `SUBSYSTEM="msr"` to set owner/group/mode.

This is largely inspired by [`nixos/modules/hardware/cpu/intel-sgx.nix`](https://github.com/NixOS/nixpkgs/blob/9c5014c5d11c92d73bb3597e1baffab3fd3f3b57/nixos/modules/hardware/cpu/intel-sgx.nix) which is a little more complex.

For background information on MSRs and the kernel module, see:
 - [Model-Specific Register (Wikipedia)](https://en.wikipedia.org/wiki/Model-specific_register)
 - [Intel MSR Tools (GitHub)](https://github.com/intel/msr-tools) (discontinued), packaged as [`msr-tools`](https://github.com/NixOS/nixpkgs/blob/9c5014c5d11c92d73bb3597e1baffab3fd3f3b57/pkgs/os-specific/linux/msr-tools/default.nix).
 - [Reading and Writing Model Specific Registers (MSRs) in Linux (Intel Article)](https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/best-practices/reading-writing-msrs-in-linux.html)
 - [`msr(4)`](https://man7.org/linux/man-pages/man4/msr.4.html)

I am planning to request the addition of packages (and respective modules) that make use of MSRs in the next weeks, so this is in preparation for:

 - [BenchExec (GitHub)](https://github.com/sosy-lab/benchexec), a framework for reliable benchmarking and resource measurement.
 - [CPU Energy Meter (GitHub)](https://github.com/sosy-lab/cpu-energy-meter), a tool for measuring energy consumption of Intel CPUs.
 - [`pqos-wrapper` (GitLab)](https://gitlab.com/sosy-lab/software/pqos-wrapper), an alternative command-line interface of [Intel PQoS](https://github.com/intel/intel-cmt-cat/tree/master/pqos) for the purpose of using it in BenchExec.

#### Security Note

Regular file permissions are not enough to actually open an msr device, the capability `CAP_SYS_RAWIO` (see [`capabilities(7)`](https://man7.org/linux/man-pages/man7/capabilities.7.html)) is required in addition (see [`msr_open` in Linux](https://github.com/torvalds/linux/blob/cbf3a2cb156a2c911d8f38d8247814b4c07f49a2/arch/x86/kernel/msr.c#L206-L222)). Further, to allow writes to these devices, the kernel parameter `msr.allow-writes=on` is required (which motivates `hardware.cpu.x86.msr.settings`).

## Things done

Tested the assertions locally, and I am using my machine with `hardware.cpu.x86.msr.enable = true`.

- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lorenzleutgeb** [comment]
@RaitoBezarius we had a brief related conversation on Matrix. The additional context you have might make this for an easy review.

File: nixos/modules/hardware/cpu/intel-msr.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  inherit (builtins) hasAttr;
+  inherit (lib) mkIf mdDoc;
+  cfg = config.hardware.cpu.intel.msr;
+  opt = options.hardware.cpu.intel.msr;
+  defaultGroup = "msr";
+  isDefaultGroup = cfg.group == defaultGroup;
+  set = "to set for devices of the `msr` kernel subsystem.";
+in
+{
+  options.hardware.cpu.intel.msr = with lib.options; {
```
Comment (RaitoBezarius) [inline nixos/modules/hardware/cpu/intel-msr.nix:16]: why is it intel specific?

Comment (lorenzleutgeb) [reply nixos/modules/hardware/cpu/intel-msr.nix:16]: Hmm... I only came across references to Intel CPUs, but you are right, there seem to be AMD CPUs that support it. Which name do you suggest? `hardware.cpu.x86.msr`?

Comment (lorenzleutgeb) [reply nixos/modules/hardware/cpu/intel-msr.nix]: Renamed to `hardware.cpu.x86.msr`.

File: nixos/doc/manual/release-notes/rl-2111.section.md [hunk 2]
```diff
@@ -581,3 +581,5 @@ In addition to numerous new and upgraded packages, this release has the followin
 - More jdk and jre versions are now exposed via `java-packages.compiler`.
 
 - The sets `haskell.packages` and `haskell.compiler` now contain for every ghc version an attribute with the minor version dropped. E.g. for `ghc8107` there also now exists `ghc810`. Those attributes point to the same compilers and packagesets but have the advantage that e.g. `ghc92` stays stable when we update from `ghc925` to `ghc926`.
+
+- The Linux kernel module `msr` (see [`msr(4)`](https://man7.org/linux/man-pages/man4/msr.4.html)), which provides an interface to read and write the model-specific registers (MSRs) of an x86 CPU, can now be configured via `hardware.cpu.x86.msr`.
```
Comment (RaitoBezarius) [inline nixos/doc/manual/release-notes/rl-2111.section.md:585]: This is 21.11 release notes, not 23.11

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
LGTM modulo release notes

File: nixos/doc/manual/release-notes/rl-2111.section.md (hunk 2 above)
Comment (lorenzleutgeb) [reply nixos/doc/manual/release-notes/rl-2111.section.md]: Ooops...

**Review by RaitoBezarius — APPROVED** [review]

**Pull request merged**